### PR TITLE
[IMP] stock_picking_invoice_link: Allow invoice lines linked with more than one stock move

### PR DESCRIPTION
**Pull request by sergio-teruel**
cc @Tecnativa TT20017

### CODE REVIEW

File: stock_picking_invoice_link/i18n/es.po [hunk 1]
```diff
@@ -89,5 +90,11 @@ msgstr "Transferir"
 msgid "View Invoice"
 msgstr "Ver factura"
 
+#. module: stock_picking_invoice_link
+#: code:addons/stock_picking_invoice_link/models/stock_move.py:28
+#, python-format
+msgid "You can not modify an invoiced stock move"
+msgstr "No puedes modificar un movimiento de almacén facturado"
```
Comment (pedrobaeza) [inline stock_picking_invoice_link/i18n/es.po]: Always "trato de usted"

```suggestion
msgstr "No puede modificar un movimiento de almacén facturado"
```

File: stock_picking_invoice_link/models/account_invoice.py [hunk 2]
```diff
@@ -19,16 +19,28 @@ class AccountInvoice(models.Model):
              "(only when the invoice has been generated from a sale order).",
     )
 
+    @api.model
+    def _refund_cleanup_lines(self, lines):
+        result = super(AccountInvoice, self)._refund_cleanup_lines(lines)
+        if self.env.context.get('mode') == 'modify':
+            for i, line in enumerate(lines):
```
Comment (pedrobaeza) [inline stock_picking_invoice_link/models/account_invoice.py:26]: Isn't there any `_prepare_refund_line` or similar for not using this weak link through line index?

File: stock_picking_invoice_link/models/stock_move.py [hunk 3]
```diff
@@ -3,15 +3,27 @@
...
         string='Invoice Line',
         copy=False,
         readonly=True,
     )
+
+    def write(self, vals):
+        if 'product_uom_qty' not in vals:
+            return super().write(vals)
+        for move in self:
+            if move.state == 'done' and move.invoice_line_ids:
```
Comment (pedrobaeza) [inline stock_picking_invoice_link/models/stock_move.py]: This is difficult to be read, having also 2 calls to super. Please use 3 levels of indentation this time.

File: stock_picking_invoice_link/models/stock_move.py [hunk 4]
```diff
@@ -3,15 +3,27 @@
...
         comodel_name='account.invoice.line',
+        relation='stock_move_invoice_line_rel',
+        column1='move_id',
+        column2='invoice_line_id',
         string='Invoice Line',
         copy=False,
         readonly=True,
     )
+
+    def write(self, vals):
```
Comment (pedrobaeza) [inline stock_picking_invoice_link/models/stock_move.py:23]: Please a docstring explaining why you are preventing this.

File: stock_picking_invoice_link/views/stock_view.xml [hunk 5]
```diff
@@ -13,12 +13,13 @@
...
-                    groups="account.group_account_invoice"
-                    attrs="{'invisible':[('invoice_ids', '=', False)]}"
-                    widget="many2many_tags"
-                    context="{'tree_view_ref': 'account.invoice_tree', 'form_view_ref': 'account.invoice_form'}"/>
+            <xpath expr="//notebook" position="inside">
+                <page name="invoice" string="Invoices" attrs="{'invisible':[('invoice_ids', '=', [])]}">
+                    <field name="invoice_ids" string="Invoice"
+                        groups="account.group_account_invoice"
+                        widget="one2many_list"
+                        context="{'tree_view_ref': 'account.invoice_tree', 'form_view_ref': 'account.invoice_form'}"/>
```
Comment (pedrobaeza) [inline stock_picking_invoice_link/views/stock_view.xml]: This is not valid when `stock_picking_link_purchase` finally arrives.

Comment (pedrobaeza) [reply stock_picking_invoice_link/views/stock_view.xml]: And why putting this on an extra tab and invoice lines in the same form?

**Review by pedrobaeza — CHANGES_REQUESTED** [review]

File: stock_picking_invoice_link/models/account_invoice.py (hunk 2 above)
Comment (sergio-teruel) [reply stock_picking_invoice_link/models/account_invoice.py:26]: I think that this ok

File: stock_picking_invoice_link/views/stock_view.xml (hunk 5 above)
Comment (sergio-teruel) [reply stock_picking_invoice_link/views/stock_view.xml]: > This is not valid when stock_picking_link_purchase finally arrives.

Ohh yea...
I thin that I can remove this and make a new smart button... but when we will have two invoice types... What view display??

> And why putting this on an extra tab and invoice lines in the same form?

I do not understand

**Comment by sergio-teruel** [comment]
@pedrobaeza some changes done!

File: stock_picking_invoice_link/views/stock_view.xml (hunk 5 above)
Comment (pedrobaeza) [reply stock_picking_invoice_link/views/stock_view.xml]: Yeah, a smart-button with a method that checks linked invoice type and return the corresponding form/tree is a good option.

File: stock_picking_invoice_link/migrations/11.0.2.0.0/post-migrate.py [hunk 6]
```diff
@@ -0,0 +1,14 @@
+# Copyright 2019 Sergio Teruel <sergio.teruel@tecnativa.com>
+# License LGPL-3.0 or later (http://www.gnu.org/licenses/lgpl).
+from openupgradelib import openupgrade
+
+
+@openupgrade.migrate()
+def migrate(env, version):
+    sql = """
```
Comment (pedrobaeza) [inline stock_picking_invoice_link/migrations/11.0.2.0.0/post-migrate.py]: There's an openupgradelib method for avoiding you to write this SQL: https://github.com/OCA/openupgradelib/blob/884331200591e098ed1a0963934660bcef8075c0/openupgradelib/openupgrade.py#L1235

File: stock_picking_invoice_link/models/stock_move.py (hunk 4 above)
Comment (pedrobaeza) [reply stock_picking_invoice_link/models/stock_move.py:23]: Reading the current docstring, shouldn't it block `qty_done` instead of `product_uom_qty`?

Comment (sergio-teruel) [reply stock_picking_invoice_link/models/stock_move.py:23]: for a picking in done state when user modifies the units... the field product_uom_qty is writed too.

**Comment by sergio-teruel** [comment]
Changes done

**Review by pedrobaeza — APPROVED** [review]

**Review by carlosdauden — APPROVED** [review]

**Comment by gustavovalverde** [comment]
This broke everything on production. This PR is using `openupgradelib` which is not in this repo requirements.

cc: @pedrobaeza

**Comment by pedrobaeza** [comment]
@gustavovalverde it shouldn't be in requirements. It's only for upgrading, not for using the module in a fresh installation. And it doesn't break anything: just revert to previous commit, or have openupgradelib in your system.

**Comment by gustavovalverde** [comment]
Pulling this changes in **running environment** does not allow the users to use pickings or invoices because of the missing fields `invoice_line_id` and `move_line_ids` which were changed from One2many to Many2one.

The views will complain as this fields are missing. And trying to upgrade the module will crash because this migration depends on `openupgradelib` which is not a requirement on this repo.

This library should be added in the requirements, as anyone who updates this source code will probably go through the same.

**Comment by pedrobaeza** [comment]
I don't agree and this has been discussed other times. Any way, just add the library and update again. I don't know why you don't have it already: it's something basic in OCA.

**Comment by gustavovalverde** [comment]
As a good practice, we're not supposed to add libraries "just in case" in production. And I'm not aware of any OCA documentation that implies this library must be in all installations which have an OCA repo in it.

**Comment by pedrobaeza** [comment]
You can contribute with this reference in the place you consider. As said, it's already discussed in other places why it's not put in external dependencies (Odoo doesn't init at all, etc), so I don't want to keep this discussion again. Take note from now on to put openupgradelib everywhere and you won't have this problem again.

**Comment by gustavovalverde** [comment]
I'll add that note here: https://github.com/OCA/odoo-community.org/blob/master/website/Contribution/CONTRIBUTING.rst

**Comment by pedrobaeza** [comment]
I don't think that place is the right one, as they are guidelines for development, not any kind of deployment guide.

**Comment by andres-pcg** [comment]
Alright @pedrobaeza, I understood the whole thing here. You have made your point and it's clear; but, where should we have that note added to? Do you have any suggestions? I know and we will for sure make a note of this. But i think there should be something shared with whole community.

**Comment by pedrobaeza** [comment]
For me, we should add openupgradelib to https://github.com/OCA/OCB/blob/12.0/requirements.txt, so that we stated that in "OCA world", that library is used with no explicit dependency declaration, the same as the rest of the libraries here, and we should add it in the source form for getting latest version:
```
git+https://github.com/OCA/openupgradelib.git@master
```

**Comment by rousseldenis** [comment]
> @gustavovalverde it shouldn't be in requirements. It's only for upgrading, not for using the module in a fresh installation. And it doesn't break anything: just revert to previous commit, or have openupgradelib in your system.

Hi @pedrobaeza. The good practice is to put it in external dependencies in manifest. If you manage your project with setuptools-odoo, it will integrate it.

But as you said, the good practice is to include it in deployments. @gustavovalverde You should have discovered that error during your integration tests.

**Comment by pedrobaeza** [comment]
@rousseldenis we had this discussion already in the past. The question is that the module doesn't need that library for a fresh install, only for migrating.

**Comment by rousseldenis** [comment]
> @rousseldenis we had this discussion already in the past. The question is that the module doesn't need that library for a fresh install, only for migrating.

Ok, but I like explicit declarations. The migration code is part of the module, so...

**Comment by pedrobaeza** [comment]
And then I talked to you about the module avoiding to init the Odoo service when introducing the external dependency without no clear clue in the logging, which it's worst, and that's why we decided this is the less bad.

**Pull request merged**